File: src/fingerprint.ts
// Fingerprint algorithm v1 — must stay in sync with:
//   web/lib/ai/fingerprint.ts
//   cli/src/mcp/fingerprint.rs
//
// Normalization steps (ORDER MATTERS for cross-language determinism):
//   1. Concatenate title + body, lowercase
//   2. Strip UUIDs (before epochs — UUIDs contain digit sequences)
//   3. Strip ISO 8601 timestamps (lowercase t)
//   4. Strip Unix epochs (10+ digits)
//   5. Strip relative times ("5 minutes ago")
//   6. Strip hex IDs (>8 chars)
//   7. Strip file paths (/foo/bar.ts)
//   8. Strip line numbers (at line 42, :42:10)
//   9. Strip URLs
//  10. Strip version numbers (v1.2.3)
//  11. Collapse whitespace, trim
//  12. SHA-256 → hex string (64 chars)

let cryptoModule: typeof import("crypto") | null = null

async function getNodeCrypto(): Promise<typeof import("crypto") | null> {
  if (cryptoModule) return cryptoModule
  try {
    cryptoModule = await import("crypto")
    return cryptoModule
  } catch {
    return null
  }
}

function normalizeErrorText(input: string): string {
  let s = input

  // 2. UUIDs (before epochs — UUIDs contain digit sequences that epoch regex would eat)
  s = s.replace(/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}/g, "<uuid>")

  // 3. ISO 8601 timestamps (lowercase t — input is already lowercased)
  s = s.replace(/\d{4}-\d{2}-\d{2}t\d{2}:\d{2}:\d{2}[^\s]*/g, "<timestamp>")

  // 4. Unix epochs (10-13 digits)
  s = s.replace(/\b\d{10,13}\b/g, "<timestamp>")

  // 5. Relative times
  s = s.replace(/\b\d+\s*(ms|seconds?|minutes?|hours?|days?)\s*ago\b/g, "<time_ago>")

  // 6. Hex IDs (>8 chars)
  s = s.replace(/\b[0-9a-f]{9,}\b/g, "<hex_id>")

  // 7. File paths
  s = s.replace(/(?:\/[\w.\-]+){2,}(?:\.\w+)?/g, "<path>")

  // 8. Line numbers
  s = s.replace(/(?:at line|line:?|:\d+:\d+)\s*\d+/g, "at line <N>")

  // 9. URLs
  s = s.replace(/https?:\/\/[^\s)]+/g, "<url>")

  // 10. Version numbers
  s = s.replace(/v?\d+\.\d+\.\d+[^\s]*/g, "<version>")

  // 11. Collapse whitespace
  return s.replace(/\s+/g, " ").trim()
}

async function sha256Hex(input: string): Promise<string> {
  // Try Node.js crypto first (fastest, synchronous)
  const nodeCrypto = await getNodeCrypto()
  if (nodeCrypto?.createHash) {
    return nodeCrypto.createHash("sha256").update(input, "utf8").digest("hex")
  }

  // Fallback: Web Crypto API (edge runtimes)
  const encoder = new TextEncoder()
  const data = encoder.encode(input)
  const hashBuffer = await crypto.subtle.digest("SHA-256", data)
  const hashArray = Array.from(new Uint8Array(hashBuffer))
  return hashArray.map((b) => b.toString(16).padStart(2, "0")).join("")
}

export async function computeErrorFingerprint(title: string, body: string): Promise<string> {
  const input = `${title}\n${body}`.toLowerCase()
  const normalized = normalizeErrorText(input)
  return sha256Hex(normalized)
}

File: rust/src/fingerprint.rs
//! Fingerprint algorithm v1 — byte-identical to:
//!
//! - `capture/src/fingerprint.ts` (Node SDK)
//! - `capture/python/.../fingerprint.py` (Python SDK)
//! - `capture/go/fingerprint.go` (Go SDK)
//! - `web/lib/ai/fingerprint.ts` (web ingest)
//! - `cli/src/mcp/fingerprint.rs` (Rust CLI)
//!
//! Steps (ORDER MATTERS):
//! 1. Concatenate `title + "\n" + body`, lowercase the whole thing.
//! 2. UUID -> `<uuid>` (must run before epoch — UUIDs contain digit runs).
//! 3. ISO 8601 timestamps (`t` is already lowercased) -> `<timestamp>`.
//! 4. Unix epochs (10-13 digits) -> `<timestamp>`.
//! 5. Relative times (`5 minutes ago`) -> `<time_ago>`.
//! 6. Hex IDs (9+ chars) -> `<hex_id>`.
//! 7. File paths (`/foo/bar.ts`) -> `<path>`.
//! 8. Line numbers (`at line 42`, `:42:10`) -> `at line <N>`.
//! 9. URLs -> `<url>`.
//! 10. Version numbers -> `<version>`.
//! 11. Collapse whitespace, trim.
//! 12. SHA-256, lowercase hex.

use sha2::{Digest, Sha256};

use crate::regex_set::regex_cache;

pub fn compute_error_fingerprint(title: &str, body: &str) -> String {
    let combined = format!("{}\n{}", title, body).to_lowercase();
    let normalized = normalize(&combined);
    let mut hasher = Sha256::new();
    hasher.update(normalized.as_bytes());
    hex::encode(hasher.finalize())
}

fn normalize(text: &str) -> String {
    let r = regex_cache();
    let mut s = text.to_string();
    s = r.uuid.replace_all(&s, "<uuid>").into_owned();
    s = r.iso_timestamp.replace_all(&s, "<timestamp>").into_owned();
    s = r.unix_epoch.replace_all(&s, "<timestamp>").into_owned();
    s = r.rel_time.replace_all(&s, "<time_ago>").into_owned();
    s = r.hex_id.replace_all(&s, "<hex_id>").into_owned();
    s = r.path.replace_all(&s, "<path>").into_owned();
    s = r.line_no.replace_all(&s, "at line <N>").into_owned();
    s = r.url.replace_all(&s, "<url>").into_owned();
    s = r.version.replace_all(&s, "<version>").into_owned();
    s = r.whitespace.replace_all(&s, " ").into_owned();
    s.trim().to_string()
}
